### PR TITLE
Fix OpenTelemetryProducer to work with the security manager.

### PR DESCRIPTION
**Pull request by rhusar**
Downstream issue in WF which still supports security manager:
https://issues.redhat.com/browse/WFLY-20043

### CODE REVIEW

**Comment by rhusar** [comment]
@bstansberry I could use your insight here, while this does fix our current issues, I am not sure whether we require this in SR OpenTelemetry or whether we can fix this in WildFly, initializing Otel early enough - is that possible @jasondlee? cc @kabir

File: implementation/cdi/src/main/java/io/smallrye/opentelemetry/implementation/cdi/OpenTelemetryProducer.java [hunk 1]
```diff
@@ -31,30 +33,72 @@
...
+        if (System.getSecurityManager() == null) {
+            garbageCollectorObservers = Cpu.registerObservers(openTelemetry);
+        } else {
+            // Requires FilePermission/RuntimePermission
+            garbageCollectorObservers = AccessController.doPrivileged(
+                    (PrivilegedAction<List<AutoCloseable>>) () -> GarbageCollector.registerObservers(openTelemetry));
+        }
+        closeables.addAll(garbageCollectorObservers);
+
+        if (System.getSecurityManager() == null) {
```
Comment (jasondlee) [inline implementation/cdi/src/main/java/io/smallrye/opentelemetry/implementation/cdi/OpenTelemetryProducer.java:93]: Can all of these SM checks be reduced to one and have all the work done together? The repeated checks makes the code a bit noisy...

**Review by jasondlee — CHANGES_REQUESTED** [review]

File: implementation/cdi/src/main/java/io/smallrye/opentelemetry/implementation/cdi/OpenTelemetryProducer.java (hunk 1 above)
Comment (jasondlee) [reply implementation/cdi/src/main/java/io/smallrye/opentelemetry/implementation/cdi/OpenTelemetryProducer.java:93]: Something like this should work, no?

```
        // Requires FilePermission/RuntimePermission/ManagementPermission
        List<AutoCloseable> cpuObservers;
        List<AutoCloseable> garbageCollectorObservers;

        if (System.getSecurityManager() == null) {
            cpuObservers = Cpu.registerObservers(openTelemetry);
            garbageCollectorObservers = Cpu.registerObservers(openTelemetry);
        } else {
            cpuObservers = AccessController
                    .doPrivileged((PrivilegedAction<List<AutoCloseable>>) () -> Cpu.registerObservers(openTelemetry));
            garbageCollectorObservers = AccessController.doPrivileged(
                    (PrivilegedAction<List<AutoCloseable>>) () -> GarbageCollector.registerObservers(openTelemetry));
        }

        closeables.addAll(cpuObservers);
        closeables.addAll(MemoryPools.registerObservers(openTelemetry));
        closeables.addAll(Threads.registerObservers(openTelemetry));
        closeables.addAll(garbageCollectorObservers);

        if (System.getSecurityManager() == null) {
            OpenTelemetryHandler.install(openTelemetry);
        } else {
            // Requires LoggingPermission
            AccessController.doPrivileged((PrivilegedAction<List<AutoCloseable>>) () -> {
                OpenTelemetryHandler.install(openTelemetry);
                return null;
            });
        }
```

Comment (rhusar) [reply implementation/cdi/src/main/java/io/smallrye/opentelemetry/implementation/cdi/OpenTelemetryProducer.java:90]: OK, if `#registerObservers(OpenTelemetry)` has no side-effects, we can don that. Done.

**Review by jasondlee — APPROVED** [review]

**Comment by bstansberry** [comment]
@rhusar I think this is ok. Dealing with SM issues at the finest granularity possible is best IMHO. I don't know enough about what it would take to do it in WF code but I expect it would be more coarse grained.

**Pull request merged**